### ISSUE TITLE
About the pretrained model and validation set in your codes

### ISSUE THREAD

**Issue by xiangyu19** [megvii-research/DPGN#8]
Thanks so much for your extraordinary work! I have 2 questions on your codes. 
1. How did you get the pretrained model? When I train from scratch without any pretrained model, the train_edge_loss is around 4 while if I used the pretrained model, it's around 1. And the testing accuracy is 66.32 for 5-way 1-shot ResNet12 (which is supposed to be 67.77)... I am a little confused about this.
2. It seems that you didn't use the validation set in your main.py since partition = 'test' in your train(). If this is true, then are you using the testing set to get the best accuracy directly? Or did you set the partition to 'val' when you were training the model?

**Comment by zilunzhang** [comment]
Hi Xiangyu,

Thanks for being interested in our work!

1.  - **No pertained model** is used in DPGN. 
     - Please note that different devices and environments could lead to slightly different results and we have listed our environment in the readme (we reproduced the results several times, and you could find the training log and checkpoint in the corresponding directories).

2. No, in fact, we didn't use the validation set in the training stage. For convenience and faster testing, we used the testing set to get the best accuracy directly. From our experiment results, it is comparable to the one uses the validation set for pre-testing.

Let us know if you have any other questions when using the DPGN.

Yours,
DPGN Team

**Comment by xiangyu19** [comment]
> Hi Xiangyu,
> 
> Thanks for being interested in our work!
> 
> 1. * **No pertained model** is used in DPGN.
>    * Please note that different devices and environments could lead to slightly different results and we have listed our environment in the readme (we reproduced the results several times, and you could find the training log and checkpoint in the corresponding directories).
> 2. No, in fact, we didn't use the validation set in the training stage. For convenience and faster testing, we used the testing set to get the best accuracy directly. From our experiment results, it is comparable to the one uses the validation set for pre-testing.
> 
> Let us know if you have any other questions when using the DPGN.
> 
> Yours,
> DPGN Team

Thanks for your immediate reply. For the first question, is that possible that it is because I set the num_workers to 4 because of the memory problem while you used 8 instead? I doubt the version of Python and CUDA might not be the reason of nearly 1% decrease in accuracy.

**Comment by xiangyu19** [comment]
The log.txt info from your best_log:
`[2020-04-19 18:06:37,757] [main] -------------configs-------------
[2020-04-19 18:06:37,758] [main] OrderedDict([('dataset_name', 'mini-imagenet'), ('num_generation', 6), ('emb_size', 128), ('backbone', 'resnet12'), ('train_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 25), ('iteration', 100000), ('lr', 0.001), ('weight_decay', 1e-05), ('dec_lr', 17000), ('dropout', 0.1), ('loss_indicator', [1, 1, 0]), ('num_queries', 1)])), ('eval_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 10), ('iteration', 1000), ('interval', 1000), ('num_queries', 1)]))])
`
and below is from my log.txt:
`[2020-06-29 20:33:31,884] [main] -------------configs-------------
[2020-06-29 20:33:31,884] [main] OrderedDict([('dataset_name', 'mini-imagenet'), ('num_generation', 6), ('num_loss_generation', 3), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2'), ('emb_size', 128), ('backbone', 'resnet12'), ('train_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 25), ('iteration', 100000), ('lr', 0.001), ('weight_decay', 1e-05), ('dec_lr', 17000), ('dropout', 0.1), ('loss_indicator', [1, 1, 0]), ('lr_adj_base', 0.1), ('num_queries', 1)])), ('eval_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 10), ('iteration', 1000), ('interval', 1000), ('num_queries', 1)]))])
`
In your configs, you don't have 
`('num_loss_generation', 3), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2')`
. Is this the reason that these parameters might not be the same?

**Comment by zilunzhang** [comment]
> The log.txt info from your best_log:
> 
> `[2020-04-19 18:06:37,757] [main] -------------configs------------- [2020-04-19 18:06:37,758] [main] OrderedDict([('dataset_name', 'mini-imagenet'), ('num_generation', 6), ('emb_size', 128), ('backbone', 'resnet12'), ('train_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 25), ('iteration', 100000), ('lr', 0.001), ('weight_decay', 1e-05), ('dec_lr', 17000), ('dropout', 0.1), ('loss_indicator', [1, 1, 0]), ('num_queries', 1)])), ('eval_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 10), ('iteration', 1000), ('interval', 1000), ('num_queries', 1)]))]) `
> 
> and below is from my log.txt:
> 
> `[2020-06-29 20:33:31,884] [main] -------------configs------------- [2020-06-29 20:33:31,884] [main] OrderedDict([('dataset_name', 'mini-imagenet'), ('num_generation', 6), ('num_loss_generation', 3), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2'), ('emb_size', 128), ('backbone', 'resnet12'), ('train_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 25), ('iteration', 100000), ('lr', 0.001), ('weight_decay', 1e-05), ('dec_lr', 17000), ('dropout', 0.1), ('loss_indicator', [1, 1, 0]), ('lr_adj_base', 0.1), ('num_queries', 1)])), ('eval_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 10), ('iteration', 1000), ('interval', 1000), ('num_queries', 1)]))]) `
> 
> In your configs, you don't have
> 
> `('num_loss_generation', 3), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2')`
> 
> Is this the reason that these parameters might not be the same?

Hi Xiangyu,

1. Do you mean num_worker in dataloader? The change of num_worker may affect the randomness in dataloader, and effect the result in some way. Due to the hardware limitation, we only tested the model in our listed environment (see Readme.md), and the results are guaranteed to be reproduced under this environment. For different environments, we think there could be some variation in the results, but we are not able to quantitatively analyze them due to the reason mentioned above.

2. No, that's definitely not the reason. For the setting of 5-1-resnet12-mini, we used ('num_loss_generation', 3), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2') all the time. The reason it does not show on the log is we didn't separate these configs to the config file when we first pushed the source code a few months ago (we only released 5-1-resnet12-mini and 5-5-resnet12-mini at that time). Some users asked for settings of convnet later, and we separate these four configs at that time.


Yours,

DPGN team

**Comment by donglongzi** [comment]
For the second question, I also think the test set will not be able to evaluate the generalization ability of the model. Because the model uses the test set to adjust the hyperparameters during the training phase. Specifically, the relevant information of the test set has been leaked to the model.
